### PR TITLE
Optimize `StyleBoxFlat.draw()`

### PR DESCRIPTION
**Pull request by MewPurPur**
I'm not great at optimizing, but I needed this... I don't know which changes contributed most to the speedup, but resizing beforehand and the unfolding of the last nested loop seem like the primary two.

```gdscript
# Example of how I tested this
func _ready() -> void:
	var sbf := StyleBoxFlat.new()
	sbf.border_color = Color.RED
	sbf.set_corner_radius_all(4)
	var t := Time.get_ticks_msec()
	for i in 1000:
		sbf.draw(get_canvas_item(), Rect2(10, 10, 60, 20))
	print(Time.get_ticks_msec() - t)
```

**Benchmarks:**

- Rounded stylebox: 205ms on master, 110ms in my PR (1.9x faster)
- Stylebox with sharp corners: 75ms on master, 65ms in my PR (1.2x faster)

On production builds, on 4.3.beta2 I get 58ms on average, on this artifact I got 39ms on average.

This speedup should make stylebox drawing faster, for example if you draw hundreds of styleboxes in the same frame. It won't affect stylebox rendering after they have been drawn.

### CODE REVIEW

**Comment by Calinou** [comment]
cc @davthedev

**Comment by AThousandShips** [comment]
I'd suggest using `ptrw` to improve iteration if we're resizing

**Comment by MewPurPur** [comment]
> I'd suggest using `ptrw` to improve iteration if we're resizing

Not too familiar with pointers, but I think I got it? I got slightly better measures, but I don't think they are statistically-significant

**Comment by MewPurPur** [comment]
Changing title as I discovered a notable speedup in drawing styleboxes in general.

**Comment by MewPurPur** [comment]
I expected that some recent optimizations, like the Vector.insert() optimization, might change the benchmark. But I still measure the same 1.9x speedup for rounded styleboxes and 1.2x for all others.

File: scene/resources/style_box_flat.cpp [hunk 1]
```diff
@@ -280,30 +263,44 @@ inline void draw_rounded_rectangle(Vector<Vector2> &verts, Vector<int> &indices,
...
 	// method either draw a ring or a filled rounded rectangle, but not both.
-	int max_inner_outer = draw_border ? 2 : 1;
-
-	for (int corner_index = 0; corner_index < 4; corner_index++) {
+	real_t quarter_arc_rad = Math_PI / 2.0;
+	Point2 style_rect_center = style_rect.get_center();
+
+	int colors_size = colors.size();
+	int verts_size = verts.size();
+	int new_verts_amount = (adapted_corner_detail + 1) * (draw_border ? 8 : 4);
```
Comment (Geometror) [inline scene/resources/style_box_flat.cpp:271]:
```suggestion
	const real_t quarter_arc_rad = Math_PI / 2.0;
	const Point2 style_rect_center = style_rect.get_center();

	const int colors_size = colors.size();
	const int verts_size = verts.size();
	const int new_verts_amount = (adapted_corner_detail + 1) * (draw_border ? 8 : 4);
```
For consistency.

**Review by Geometror — APPROVED** [review]
Haven't benchmarked myself, but I believe you :) Great work!
While some aspects slightly impact readability, the performance boost is significant enough to justify the trade-off I think. I've played around with `StyleBoxFlat`s for quite a bit and haven't observed any regressions.

File: scene/resources/style_box_flat.cpp [hunk 2]
```diff
@@ -280,30 +263,44 @@ inline void draw_rounded_rectangle(Vector<Vector2> &verts, Vector<int> &indices,
...
+			}
 
-				const real_t x = radius * (real_t)cos((corner_index + detail / (double)adapted_corner_detail) * (Math_TAU / 4.0) + Math_PI) + corner_point.x;
-				const real_t y = radius * (real_t)sin((corner_index + detail / (double)adapted_corner_detail) * (Math_TAU / 4.0) + Math_PI) + corner_point.y;
-				const float x_skew = -skew.x * (y - style_rect.get_center().y);
-				const float y_skew = -skew.y * (x - style_rect.get_center().x);
-				verts.push_back(Vector2(x + x_skew, y + y_skew));
-				colors.push_back(color);
+			const real_t pt_angle = (corner_idx + detail / (double)adapted_corner_detail) * quarter_arc_rad + Math_PI;
+			const real_t angle_cosine = cos(pt_angle);
```
Comment (AThousandShips) [inline scene/resources/style_box_flat.cpp:285]:
```suggestion
			const real_t angle_cosine = Math::cos(pt_angle);
```

File: scene/resources/style_box_flat.cpp [hunk 3]
```diff
@@ -280,30 +263,44 @@ inline void draw_rounded_rectangle(Vector<Vector2> &verts, Vector<int> &indices,
...
+			const real_t angle_sine = sin(pt_angle);
```
Comment (AThousandShips) [inline scene/resources/style_box_flat.cpp:286]:
```suggestion
			const real_t angle_sine = Math::sin(pt_angle);
```

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Belated thanks! And oops I merged by mistake, I intended to wait for style fixes first. Can be done in a follow-up :)

**Pull request merged**